cpp/src/detail/shuffle_vertex_pairs.cu

```diff
@@ -80,58 +82,93 @@ shuffle_vertex_pairs_by_gpu_id_impl(
   std::optional<rmm::device_uvector<edge_t>> rx_edge_ids{std::nullopt};
   std::optional<rmm::device_uvector<edge_type_t>> rx_edge_types{std::nullopt};
 
+  // FIXME: Consider a generic function that takes a value tuple of optionals
+  //        to eliminate this complexity
```

Comment:
Yeah... I have thought about this in the past, but not sure this is possible as long as we want to explicitly instantiate for different type combinations.

Updating `groupby_and_count` or creating a wrapper function for `groupby_and_count` to take optional objects and dynamically evaluate them sounds more like pushing the complexity to one lower level, but this sounds like an overkill at this moment unless this logic gets reused in many different places (and this function is already a pretty low level detail space function).

Reply:
Perhaps overkill at the moment.  My concern is that as properties proliferate codes like this will become much more complex.  But I agree for now it's fine to defer this notion (that's why I made it a comment).